### ISSUE TITLE
Rate limit violations in v0.11.1 alpha builds (BUG REPORTS NEEDED)

### ISSUE THREAD

**Issue by gerwaric** [gerwaric/acquisition#50]
GGG says that at least one user has found themselves blocked by cloudflare for sending too many network requests.

I've never seen this before, and it's something a little different from the usual rate limit violation, so if you've experienced this issue, please say something here; your help figuring this issue out would be greatly appreciated!

**Comment by Clif4D** [comment]
That may have been me, happened yesterday (~20 hours ago).
Was switching to 0.11.1-alpha.11, and after fixing the MSVC startup problem (deleted the entire Program Files/acquisition folder and reinstalled), I could log in with OAuth.
I then moved the user data (from C:\Users\"USERNAME"\AppData\Local\acquisition  to  C:\Users\"USERNAME"\AppData\Local\acquisition_old), and started again, to get a clean dataset.
Was met with an empty tab list, but after clicking "Refresh all tabs", i got a slew of error messages, some about tabs being unable to be parsed (color code "ff" or something like that), and some about "rate limiting not indicated" or similar.
I was then rate banned from pathofexile.com via Cloudflare (returned error code 1015).

Will note that I have a huge amount of stash tabs in standard (2882), which may have been part of the problem? I have loaded the entire stash with the 0.10.x release without problems some months ago.

The log file for the 0.11.1-alpha.11 data is fairly short, it may have been overwritten or not committed to disk when the rate limit hit.
Have attached the log file from the old data set as well, maybe that is of some use.

If there's any other data you need, or something you want me to try, let me know.

[log.txt](https://github.com/user-attachments/files/16373532/log.txt)
[log_old.txt](https://github.com/user-attachments/files/16373556/log_old.txt)

**Comment by gerwaric** [comment]
Thank you for that excellent writeup. I don't know how long it will take to identify or fix this issue, but the level of detail is super helpful. I may need more information later, so I'll let you know.

In the meantime, I have two questions for you:
1. Fingers crosssed, are you still able to use v0.10?
2. At some point, would you be willing to "donate" some of your stash tabs as a set of test data?

I'm asking that 2nd question because you probably havelegacy items and tabs that I don't beacuse my account has only existed since 3.10. Many people using acquisition have much older accounts, and GGG has changed their data structures over time, so I'd love test cases with legacy items and tabs.

If you're willing, I would still have to do work to help you extract the right json objects from your data files. The reason is that acquisition's data files include additional information that should excluded from test cases, such as your account name. (If this sounds like too much of a bother--or you just don't like the idea of sharing your data--that's fine as well. People have been good about reporting issues here, and I've just implemented automatic crash reporting, so I can work from that.)

Next, here's a brief description of what happened:

From what I understand after feedback from GGG combined with your writeup, it makes sense the are short. What happened is that acquisition uses something called HEAD requests the very first time an API endpoint is called. A HEAD request doesn't return any actual data, but it does include the status of the rate limit policies that apply to that endpoint, which acquisition needs to avoid rate limit violations at startup. A HEAD request should only be sent once for each endpoint--e.g. once for the endpoint that returns the list of stash tabs, once for the endpoint that returns the list of characters, once for the endpoint that returns a specific stash tab, etc.

The problem was that somehow acquisition was sending hundreds or thousands of HEAD requests for the same endpoints over and over, which should never be needed, and which should never happen. There is no official rate-limit policy from GGG that applies to HEAD requests, but it was enough to trigger Cloudflare to block you.

Hopefully that makes some sense. I'm happy to answer any questions you have to the best of my ability.

Finally, I apologize for the inconvenience of getting your account blocked, especially with 3.25 launching tomorrow. Please bear with me and I'll try to get the underlying problem with HEAD requests fixed... or at least add extra protections against making too many requests.

Thanks, and good luck with 3.25 if you're league starting tomorrow!

**Comment by gerwaric** [comment]
Ok, I have an update:

https://github.com/gerwaric/acquisition/releases/tag/v0.11.1-alpha.12

This update doesn't fix the root cause, but it has safeguards that prevent the cascading failure that led to a flood of network requests. I highly recommend installing this update, and if there are issues with it, I will pause my league start to help.

**Comment by Clif4D** [comment]
>> Fingers crosssed, are you still able to use v0.10?
Haven't tested 0.10 again, but just had a test run of the new alpha and so far it seems to be working as expected (Though it will be ~14 hours before all the tabs are fetched).

On a related note, is there a way to fetch/update the list of tabs without getting all the content, which is quicker? Because then I could just select a subsection of the tabs to actually update the content for.

>> At some point, would you be willing to "donate" some of your stash tabs as a set of test data?

Certainly, I have a lot of old and weird items, so if that can in any way help make this application more robust, I'd be happy to help out.

>> Finally, I apologize for the inconvenience of getting your account blocked, especially with 3.25 launching tomorrow. Please bear with me and I'll try to get the underlying problem with HEAD requests fixed... or at least add extra protections against making too many requests.

No worries, not even NASA can account for all possible bugs that can happen in their software, I'm just glad you're still working on and updating the application, since playing standard without it is not as much fun, especially after GGG removed the public listing of remove-only tabs (though understandable).

>> Thanks, and good luck with 3.25 if you're league starting tomorrow!

Thanks, and you too - not sure I'll have much time to play this weekend, but will definitely try it out over the coming weeks.

**Comment by gerwaric** [comment]
I know these are famour last works, but it shouldn't be hard to refresh just the list of tabs. I'll put that on the feature list for 0.12 along with data export.

The ability to get items in remove only tabs listed on the trade site is one of my favorite features of acquisition, but it's also 
 super obscure and only matters in Standard.

**Comment by gerwaric** [comment]
I'm closing this issue since I think it's a duplicate of https://github.com/gerwaric/acquisition/issues/51 now that other people have reported the same issue.

**Comment by gerwaric** [comment]
I'm closing this issue since I think it's a duplicate of https://github.com/gerwaric/acquisition/issues/51 now that other people have reported the same issue.

**Comment by gerwaric** [comment]
**WARNING: I just learned that this bug hasn't been fixed yet. https://github.com/gerwaric/acquisition/issues/56**

With apologies, I recommend rolling back to [v0.11.0](https://github.com/gerwaric/acquisition/releases/tag/v0.11.0) or--if you are more conservative--not using acquisition until there's a newer release.
